clusterloader2/pkg/measurement/util/controlled_pods_indexer.go

```diff
@@ -40,8 +40,191 @@ type ControlledPodsIndexer struct {
+// i.e. how many pods it owns exist, whether the RS object itself exists
+// and its latest known owner's UID.
+type ReplicaSetState struct {
+	NumPods int
+	Exists  bool
+	Owner   types.UID
+}
+
+// UIDSet is a collection of ReplicaSet objects UIDs.
+type UIDSet map[types.UID]bool
+
+// NewControlledPodsIndexer creates a new ControlledPodsIndexer instance.
+func NewControlledPodsIndexer(podsInformer coreinformers.PodInformer, rsInformer appsinformers.ReplicaSetInformer) (*ControlledPodsIndexer, error) {
+	if err := podsInformer.Informer().AddIndexers(cache.Indexers{controllerUIDIndex: controllerUIDIndexFunc}); err != nil {
+		return nil, fmt.Errorf("failed to register indexer: %w", err)
+	}
+
+	cpi := &ControlledPodsIndexer{
+		podsIndexer:           podsInformer.Informer().GetIndexer(),
+		podsSynced:            podsInformer.Informer().HasSynced,
+		rsUIDToState:          make(map[types.UID]*ReplicaSetState),
+		deploymentUIDToRsUIDs: make(map[types.UID]UIDSet),
+	}
+
+	podsInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+				rsUID, kind := cpi.getControllerInfo(obj)
+				if rsUID == "" || kind != "ReplicaSet" {
+					return
+				}
+				cpi.updatePodRefCountLocked(rsUID, 1)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				oldRsUID, oldKind := cpi.getControllerInfo(oldObj)
+				newRsUID, newKind := cpi.getControllerInfo(newObj)
+				if oldRsUID == newRsUID {
+					return
+				}
+				if oldKind == "ReplicaSet" {
+					cpi.updatePodRefCountLocked(oldRsUID, -1)
+				}
+				if newKind == "ReplicaSet" {
+					cpi.updatePodRefCountLocked(newRsUID, 1)
+				}
+				if oldKind == "ReplicaSet" {
+					cpi.clearRSDataIfPossibleLocked(oldRsUID)
+				}
+			},
+			DeleteFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				rsUID, kind := cpi.getControllerInfo(obj)
+				if rsUID == "" || kind != "ReplicaSet" {
+					return
+				}
+				cpi.updatePodRefCountLocked(rsUID, -1)
+				cpi.clearRSDataIfPossibleLocked(rsUID)
+			},
+		},
+	)
+	rsInformer.Informer().AddEventHandler(
+		cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				deploymentUID, kind := cpi.getControllerInfo(obj)
+				if deploymentUID == "" || kind != "Deployment" {
+					return
+				}
+				cpi.handleIncomingRSAddEventLocked(obj)
+			},
+			UpdateFunc: func(oldObj, newObj interface{}) {
+				cpi.UIDLock.Lock()
+				defer cpi.UIDLock.Unlock()
+
+				oldDeploymentUID, oldKind := cpi.getControllerInfo(oldObj)
+				newDeploymentUID, newKind := cpi.getControllerInfo(newObj)
+				if oldDeploymentUID == newDeploymentUID {
+					return
+				}
+				if oldKind == "Deployment" {
+					cpi.handleIncomingRSDeleteEventLocked(oldObj)
```

Comment:
If the oldDeploymentUID != newDeploymentUID, you have to remove rs from deploymentUIDToRsUIDs[oldDeploymentUID].